File: pages/20260824170700_holding_topology_passing/analysis/julia/generate.jl
include("PassingAnalysis.jl")
using .PassingAnalysis

const OUTPUT_PATH = joinpath(@__DIR__, "..", "results", "occupancy_julia.csv")

function share_text(shares)
    return join(format_rational.(shares), " ")
end

function format_rational(value)
    rational = Rational{Int}(value)
    return "$(numerator(rational))/$(denominator(rational))"
end

function schedule_for(fixture)
    if fixture.grammar == "4hs"
        return schedule_four_hand(fixture.notation)
    end
    return schedule_passing(fixture.notation)
end

function metric_row(fixture)
    schedule = schedule_for(fixture)
    occupancy = occupancy_shares(schedule; dwell_ratio=DWELL_RATIO)
    bodies = body_occupancy_shares(schedule; dwell_ratio=DWELL_RATIO)
    mean_q = sum(share * (held - 1) for (held, share) in enumerate(occupancy.occupancy_shares))
    pass_count = count(event -> event.pass, schedule.cycle_tosses)
    self_count = count(event -> event.kind != "empty" && !event.pass, schedule.cycle_tosses)
    return [
        fixture.name, fixture.notation, fixture.grammar, string(schedule.hand_period),
        string(schedule.cycle_length), string(schedule.period), string(schedule.ball_count),
        string(schedule.hand_count), format_rational(BEAT_SECONDS), format_rational(DWELL_RATIO),
        format_rational(occupancy.p_alpha), format_rational(occupancy.p_polymorphy),
        format_rational(occupancy.p_kappa), share_text(occupancy.occupancy_shares),
        format_rational(mean_q), string(pass_count), string(self_count),
        join(share_text.(bodies), " | "),
    ]
end

function write_occupancy_csv(path)
    mkpath(dirname(path))
    open(path, "w") do io
        println(io, join([
            "scenario", "notation", "grammar", "hand_period", "cycle_length_beats",
            "notation_period_beats", "object_count", "hand_count", "beat_seconds", "dwell_ratio",
            "p_alpha", "p_polymorphy", "p_kappa", "occupancy_shares", "mean_q", "pass_count",
            "self_count", "body_occupancy_shares",
        ], ","))
        for fixture in FIXTURES
            println(io, join(metric_row(fixture), ","))
        end
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    write_occupancy_csv(OUTPUT_PATH)
end

File: pages/20260824170700_holding_topology_passing/analysis/ruby/generate.rb
require "csv"
require "fileutils"
require_relative "passing_schedule"
require_relative "passing_four_hand"
require_relative "passing_occupancy"
require_relative "passing_fixtures"

module PassingAnalysis
  module Generate
    module_function

    COLUMNS = %w[
      scenario notation grammar hand_period cycle_length_beats notation_period_beats object_count
      hand_count beat_seconds dwell_ratio p_alpha p_polymorphy p_kappa occupancy_shares mean_q
      pass_count self_count body_occupancy_shares
    ].freeze

    def rows
      FIXTURES.map do |fixture|
        schedule = schedule_for(fixture)
        occupancy = Occupancy.shares(schedule, dwell_ratio: DWELL_RATIO)
        bodies = Occupancy.body_shares(schedule, dwell_ratio: DWELL_RATIO)
        shares = occupancy.fetch(:occupancy_shares)
        mean_q = shares.each_with_index.sum(0.to_r) { |share, held| share * held }
        {
          scenario: fixture.fetch(:name),
          notation: fixture.fetch(:notation),
          grammar: fixture.fetch(:grammar),
          hand_period: schedule.hand_period.to_s,
          cycle_length_beats: schedule.cycle_length,
          notation_period_beats: schedule.period,
          object_count: schedule.ball_count,
          hand_count: schedule.hand_count,
          beat_seconds: format_rational(BEAT_SECONDS),
          dwell_ratio: format_rational(DWELL_RATIO),
          p_alpha: format_rational(occupancy.fetch(:p_alpha)),
          p_polymorphy: format_rational(occupancy.fetch(:p_polymorphy)),
          p_kappa: format_rational(occupancy.fetch(:p_kappa)),
          occupancy_shares: format_shares(shares),
          mean_q: format_rational(mean_q),
          pass_count: schedule.cycle_tosses.count(&:pass),
          self_count: schedule.cycle_tosses.count { |event| event.kind != "empty" && !event.pass },
          body_occupancy_shares: bodies.map { |body| format_shares(body) }.join(" | ")
        }
      end
    end

    def schedule_for(fixture)
      if fixture.fetch(:grammar) == "4hs"
        FourHand.schedule(fixture.fetch(:notation))
      else
        Scheduler.schedule(fixture.fetch(:notation))
      end
    end

    def format_rational(value)
      rational = Rational(value)
      "#{rational.numerator}/#{rational.denominator}"
    end

    def format_shares(shares)
      shares.map { |share| format_rational(share) }.join(" ")
    end

    def write_csv(path)
      FileUtils.mkdir_p(File.dirname(path))
      CSV.open(path, "w") do |csv|
        csv << COLUMNS
        rows.each { |row| csv << COLUMNS.map { |column| row.fetch(column.to_sym) } }
      end
    end
  end
end

if $PROGRAM_NAME == __FILE__
  PassingAnalysis::Generate.write_csv(
    File.expand_path("../results/occupancy_ruby.csv", __dir__)
  )
end
